zebra-state/src/service/read/tree.rs

```diff
@@ -87,25 +136,71 @@ where
         .or_else(|| db.orchard_tree_by_hash_or_height(hash_or_height))
 }
 
-/// Returns the Orchard [`NoteCommitmentSubtree`] specified by an
-/// index, if it exists in the non-finalized `chain` or finalized `db`.
-#[allow(unused)]
-pub fn orchard_subtree<C>(
+/// Returns a list of Orchard [`NoteCommitmentSubtree`]s starting at `start_index`.
+/// If `limit` is provided, the list is limited to `limit` entries.
+///
+/// If there is no subtree at `start_index` in the non-finalized `chain` or finalized `db`,
+/// the returned list is empty. Otherwise, subtrees are continuous and consistent up to the tip.
+///
+/// There is no API for retrieving single subtrees, because it can accidentally be used to create
+/// an inconsistent list of subtrees.
+pub fn orchard_subtrees<C>(
     chain: Option<C>,
     db: &ZebraDb,
-    index: NoteCommitmentSubtreeIndex,
-) -> Option<Arc<NoteCommitmentSubtree<orchard::tree::Node>>>
+    start_index: NoteCommitmentSubtreeIndex,
+    limit: Option<NoteCommitmentSubtreeIndex>,
+) -> BTreeMap<NoteCommitmentSubtreeIndex, NoteCommitmentSubtreeData<orchard::tree::Node>>
 where
     C: AsRef<Chain>,
 {
     // # Correctness
     //
-    // Since orchard treestates are the same in the finalized and non-finalized
-    // state, we check the most efficient alternative first. (`chain` is always
-    // in memory, but `db` stores blocks on disk, with a memory cache.)
-    chain
-        .and_then(|chain| chain.as_ref().orchard_subtree_by_index(index))
-        .or_else(|| db.orchard_subtree_by_index(index))
+    // After `chain` was cloned, the StateService can commit additional blocks to the finalized
+    // state `db`. Usually, the subtrees of these blocks are consistent. But if the `chain` is
+    // a different fork to `db`, then the trees can be inconsistent.
+    //
+    // In that case, we ignore all the trees in `chain` after the first inconsistent tree,
+    // because we know they will be inconsistent as well. (It is cryptographically impossible
+    // for tree roots to be equal once the leaves have diverged.)
```

Comment:
I would promote this comment to the docs for the whole function so that it's clear what it returns if there are different subtrees in the provided `db` and `chain`.